File: agent/src/narrator/content.py
import json

from narrator.config import LIBRARY_DIR


def load_story(collection: str, story_id: str) -> dict:
    path = (LIBRARY_DIR / collection / f"{story_id}.json").resolve()
    if not path.is_relative_to(LIBRARY_DIR.resolve()):
        raise ValueError(f"story is outside the library: {collection}/{story_id}")
    return json.loads(path.read_text())


def load_voice(voice_id: str) -> dict:
    voices = json.loads((LIBRARY_DIR / "voices.json").read_text())
    return {voice["id"]: voice for voice in voices}[voice_id]

File: web/lib/content.ts
import { readFile } from "node:fs/promises";
import path from "node:path";

export type Story = {
  id: string;
  title: string;
  blurb: string;
  script: string[];
};

export type Voice = {
  id: string;
  name: string;
  elevenLabsId: string;
};

const LIBRARY_DIR = path.join(process.cwd(), "..", "library");
const NAME = /^[A-Za-z0-9_-]+$/;

export async function getVoices(): Promise<Voice[]> {
  const raw = await readFile(path.join(LIBRARY_DIR, "voices.json"), "utf8");
  return JSON.parse(raw) as Voice[];
}

export async function getStory(
  collection: string,
  storyId: string,
): Promise<Story | null> {
  if (!NAME.test(collection) || !NAME.test(storyId)) return null;
  try {
    const raw = await readFile(
      path.join(LIBRARY_DIR, collection, `${storyId}.json`),
      "utf8",
    );
    return JSON.parse(raw) as Story;
  } catch {
    return null;
  }
}
